### PR TITLE
DO NOT MERGE: EE 62

### PR DESCRIPTION
**Pull request by vinnicombe**

### CODE REVIEW

File: app/viewmodels/CheckboxOption.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package viewmodels
+
+final case class CheckboxOption(value: String, messageKey: String)
```
Comment (TheNickWilson) [inline app/viewmodels/CheckboxOption.scala]: You can probably simply reuse the RadioOption view model instead of creating another (perhaps renaming it to something more general)?

File: migrations/applied_migrations/Checkbox.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+echo "POST       /changeCheckbox                  controllers.CheckboxController.onSubmit(mode: Mode = CheckMode)" >> ../conf/app.routes
+
+echo "Adding messages to conf.messages"
+echo "" >> ../conf/messages.en
+echo "checkbox.title = checkbox" >> ../conf/messages.en
+echo "checkbox.heading = checkbox" >> ../conf/messages.en
+echo "checkbox.option1 = option1" >> ../conf/messages.en
+echo "checkbox.option2 = option2" >> ../conf/messages.en
+echo "checkbox.checkYourAnswersLabel = checkbox" >> ../conf/messages.en
+echo "checkbox.error.required = Select checkbox" >> ../conf/messages.en
```
Comment (TheNickWilson) [inline migrations/applied_migrations/Checkbox.sh]: These don't quite match the model you generated (e.g. you're missing options 3 and 4)

File: migrations/applied_migrations/Checkbox.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    next }1' ../test/generators/UserDataGenerator.scala > tmp && mv tmp ../test/generators/UserDataGenerator.scala
+
+echo "Adding helper method to CheckYourAnswersHelper"
+awk '/class/ {\
+     print;\
+     print "";\
+     print "  def checkbox: Option[AnswerRow] = userAnswers.get(CheckboxPage) map {";\
+     print "    x => AnswerRow(\"checkbox.checkYourAnswersLabel\", s\"checkbox.$x\", true, routes.CheckboxController.onPageLoad(CheckMode).url)";\
+     print "  }";\
+     next }1' ../app/utils/CheckYourAnswersHelper.scala > tmp && mv tmp ../app/utils/CheckYourAnswersHelper.scala
```
Comment (TheNickWilson) [inline migrations/applied_migrations/Checkbox.sh]: This doesn't match the implementation in CheckYourAnswersHelper

File: test/forms/behaviours/CheckboxFieldBehaviours.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      val data = Map(
+        "value[0]" -> invalidValue
+      )
+      form.bind(data).errors should contain(FormError("value[0]", invalid))
+    }
+  }
+
+  def aMandatoryCheckboxForm(required: String, args: Any*): Unit = {
+
+    "fail to bind when no answers are selected" in {
```
Comment (TheNickWilson) [inline test/forms/behaviours/CheckboxFieldBehaviours.scala]: For completeness we could add another test that validates the key being present with an empty value (this test is for the key itself being missing)

File: test/generators/ModelGenerators.scala [hunk 5]
```diff
@@ -16,9 +16,14 @@
...
+import models.Checkbox.Option1
 import models._
-import org.scalacheck.Arbitrary.arbitrary
 import org.scalacheck.{Arbitrary, Gen}
 
 trait ModelGenerators {
+
+  implicit lazy val arbitraryCheckbox: Arbitrary[Checkbox] =
+    Arbitrary {
+      Gen.oneOf(Seq(Option1))
```
Comment (TheNickWilson) [inline test/generators/ModelGenerators.scala]: Instead of hardcoding to Option1, this could be something like Gen.oneOf(Checkbox.values.toSeq)

File: test/generators/UserDataGenerator.scala [hunk 6]
```diff
@@ -26,6 +26,8 @@ trait UserDataGenerator {
   self: Generators =>
 
   val generators: Seq[Gen[(Page, JsValue)]] =
+    arbitrary[(CheckboxPage.type, JsValue)] ::
+    arbitrary[(CheckboxPage.type, JsValue)] ::
```
Comment (TheNickWilson) [inline test/generators/UserDataGenerator.scala]: This doesn't need to be duplicated

File: test/models/CheckboxSpec.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import play.api.libs.json._
+import org.scalacheck.Arbitrary.arbitrary
+
+class CheckboxSpec extends WordSpec with MustMatchers with PropertyChecks with OptionValues {
+
+  "Checkbox" must {
+
+    "deserialise valid values" in {
+
+      val gen= Gen.oneOf(Checkbox.values.toSeq)
```
Comment (TheNickWilson) [inline test/models/CheckboxSpec.scala]: Could use arbitrary[Checkbox] here (and on line 53)
